amino.py

```diff
@@ -102,60 +136,41 @@ def _d2_bin(self, x, y):
     # Checks if distance has been computed before, otherwise computes distance
     def distance(self, OP1: OrderParameter, OP2: OrderParameter) -> float:
         """Returns the mutual information distance between two OPs.
-        
+
         Parameters
         ----------
         OP1 : OrderParameter
             The first order parameter for distance calculation.
-            
+
         OP2 : OrderParameter
             The second order parameter for distance calculation.
-            
+
         Returns
         -------
         float
             The mutual information distance.
-            
+
         """
 
-        index1 = f"{OP1.name} {OP2.name}"
-        index2 = f"{OP2.name} {OP1.name}"
+        idx = frozenset((OP1.name, OP2.name))
+        try:
+            d = self.memo[idx]
+        except KeyError:
+            raise ValueError(f"Distance between {OP1.name} and {OP2.name} not found.")
 
-        if index1 in self.memo:
-            return self.memo[index1]
-        elif index2 in self.memo:
-            return self.memo[index2]
+        return d
 
-        d = self._distance_kernel(OP1, OP2)
-        self.memo[index1] = d
+    def _distance_kernel(self, i, j) -> float:
```

Comment:
Done

Reply:
This `_distance_kernel` function was supposed to be a private function by design. I tried to separate this and use `jit` to accelerate this kernel function but it was unsuccessful (very little performance improvement and big overhead). The performance bottleneck was the KDE instead of the mutual information calculation. That part is likely already a C code. The final solution is just to parallelize the initialization as you see now.